src/storage/src/hummock/block_cache.rs

```diff
@@ -74,29 +84,103 @@ impl Deref for BlockHolder {
 unsafe impl Send for BlockHolder {}
 unsafe impl Sync for BlockHolder {}
 
+impl TieredCacheKey for (HummockSstableId, u64) {
+    fn encoded_len() -> usize {
+        16
+    }
+
+    fn encode(&self, mut buf: &mut [u8]) {
+        buf.put_u64(self.0);
+        buf.put_u64(self.1);
+    }
+
+    fn decode(mut buf: &[u8]) -> Self {
+        let sst_id = buf.get_u64();
+        let block_idx = buf.get_u64();
+        (sst_id, block_idx)
+    }
+}
+
+impl TieredCacheValue for Box<Block> {
+    fn len(&self) -> usize {
+        Block::len(self)
+    }
+
+    fn encoded_len(&self) -> usize {
+        self.len()
+    }
+
+    fn encode(&self, mut buf: &mut [u8]) {
+        buf.put_slice(self.raw_data());
+    }
+
+    fn decode(buf: Vec<u8>) -> Self {
+        Box::new(Block::decode_from_raw(Bytes::from(buf)))
+    }
+}
+
+pub struct MemoryBlockCacheEventListener {
+    tiered_cache: TieredCache<(HummockSstableId, u64), Box<Block>>,
+}
+
+impl LruCacheEventListener for MemoryBlockCacheEventListener {
+    type K = (HummockSstableId, u64);
+    type T = Box<Block>;
+
+    fn on_release(&self, key: Self::K, value: Self::T) {
+        // TODO(MrCroxx): handle error?
+        self.tiered_cache.insert(key, value).unwrap();
+    }
+}
+
 #[derive(Clone)]
 pub struct BlockCache {
     inner: Arc<LruCache<(HummockSstableId, u64), Box<Block>>>,
+
+    tiered_cache: TieredCache<(HummockSstableId, u64), Box<Block>>,
 }
 
 impl BlockCache {
-    pub fn new(capacity: usize, mut max_shard_bits: usize) -> Self {
+    pub fn new(
+        capacity: usize,
+        mut max_shard_bits: usize,
+        tiered_cache: TieredCache<(HummockSstableId, u64), Box<Block>>,
+    ) -> Self {
         if capacity == 0 {
             panic!("block cache capacity == 0");
         }
         while (capacity >> max_shard_bits) < MIN_BUFFER_SIZE_PER_SHARD && max_shard_bits > 0 {
             max_shard_bits -= 1;
         }
-        let cache = LruCache::new(max_shard_bits, capacity);
+
+        let listener = Arc::new(MemoryBlockCacheEventListener {
+            tiered_cache: tiered_cache.clone(),
+        });
+
+        let cache = LruCache::with_event_listener(max_shard_bits, capacity, listener);
+
         Self {
             inner: Arc::new(cache),
+
+            tiered_cache,
         }
     }
 
-    pub fn get(&self, sst_id: HummockSstableId, block_idx: u64) -> Option<BlockHolder> {
-        self.inner
+    pub async fn get(&self, sst_id: HummockSstableId, block_idx: u64) -> Option<BlockHolder> {
```

Comment:
Making `get` to be async may cause non-negligible overhead in the case we have in-memory block cache hit. I am thinking whether we should make `TieredCache` part of `SstableStore` instead of part of `BlockCache`.

Reply:
have the same concerns with patrick, get is an extremely frequent function call. Inspired by iterator testing by wenym1, `async` may cause non-negligible overhead (even if no scheduling is expected to occur)

Reply:
> Making `get` to be async may cause non-negligible overhead in the case we have in-memory block cache hit. I am thinking whether we should make `TieredCache` part of `SstableStore` instead of part of `BlockCache`.

Agreed. Then should we decouple wait queue and `LruCache`? I think there's the same problem.

Reply:
> Then should we decouple wait queue and LruCache?

We can leave it for now, combing wait queue and lru cache can avoid an additional mutex.